cachi2/core/package_managers/rpm/main.py

```diff
@@ -0,0 +1,301 @@
+            lockfile.metadata[dest] = {
+                "url": pkg.url,
+                "size": pkg.size,
+                "checksum": pkg.checksum,
+            }
+            Path.mkdir(dest.parent, parents=True, exist_ok=True)
+
+        for pkg in arch.source:
+            repoid = lockfile.internal_source_repoid if pkg.repoid is None else pkg.repoid
+            dest = output_dir.joinpath(arch.arch, repoid, Path(pkg.url).name)
+            files[pkg.url] = str(dest)
+            lockfile.metadata[dest] = {
+                "url": pkg.url,
+                "size": pkg.size,
+                "checksum": pkg.checksum,
+            }
+            Path.mkdir(dest.parent, parents=True, exist_ok=True)
+
+        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+
+def _verify_downloaded(lockfile: RedhatRpmsLock) -> None:
+    """Use metadata structure with file sizes and checksums for verification \
+    of downloaded packages and sources."""
+    log.debug("Verification of downloaded files has started.")
+
+    def raise_exception(message: str) -> None:
+        raise PackageRejected(
+            f"Some RPM packages or sources weren't verified after being downloaded: '{message}'",
+            solution=(
+                "Check the source of the data or check the corresponding metadata "
+                "in the lockfile (size, checksum)."
+            ),
+        )
+
+    # check file size and checksum of downloaded files
+    for file_path, file_metadata in lockfile.metadata.items():
+        # size is optional
+        if file_metadata["size"] is not None:
+            if file_path.stat().st_size != file_metadata["size"]:
+                raise_exception(f"Unexpected file size of '{file_path}' != {file_metadata['size']}")
+
+        # checksum is optional
+        if file_metadata["checksum"] is not None:
+            alg, digest = file_metadata["checksum"].split(":")
+            method = getattr(hashlib, alg.lower(), None)
+            if method is not None:
+                h = method(usedforsecurity=False)
+            else:
+                raise_exception(f"Unsupported hashing algorithm '{alg}' for '{file_path}'")
+            with open(file_path, "rb") as f:
+                for chunk in iter(lambda: f.read(READ_CHUNK), b""):
+                    h.update(chunk)
+            if digest != h.hexdigest():
+                raise_exception(f"Unmatched checksum of '{file_path}' != '{digest}'")
+
+
+def _generate_sbom_components(files_metadata: dict[Path, Any]) -> list[Component]:
+    """Fill the component list with the package records."""
+    components: list[Component] = []
+    for file_path, file_metadata in files_metadata.items():
+        query_format = (
+            # all nvra macros should be present/mandatory in RPM
+            "%{NAME}\n"
+            "%{VERSION}\n"
+            "%{RELEASE}\n"
+            "%{ARCH}\n"
+            # vendor and epoch are optional in RPM file and in PURL as well
```

Comment:
nitpick: can we optically split these lines that have a commentary attached?

Reply:
I did it, but later I found that putting simple spaces doesn't make _black_ happy. Converting `query_format` to a `list` doesn't help either. Other approaches I am thinking about are more complicated.

Reply:
(heavy sigh - black)...